Source/ACE/Network/NetworkSession.cs

```diff
@@ -164,59 +191,149 @@ private void ProcessPacket(ClientPacket packet)
+                    log.DebugFormat("[{0}] Adding fragment {1} to existing buffer. Buffer at {2} of {3}", session.Account, fragment.Header.Sequence, buffer.Count, buffer.TotalFragments);
+                    buffer.AddFragment(fragment);
+                    if (buffer.Complete)
                     {
-                        buffer.AddFragment(fragment);
-                        if (buffer.Complete)
-                        {
-                            ClientMessage message = buffer.GetMessage();
-                            InboundMessageManager.HandleClientMessage(message, session);
-                            partialFragments.Remove(fragment.Header.Sequence);
-                        }
-                    }
-                    else
-                    {
-                        partialFragments.Add(fragment.Header.Sequence, new MessageBuffer(fragment.Header.Count));
+                        // The buffer is complete, so we can go ahead and handle
+                        log.DebugFormat("[{0}] Buffer {1} is complete", buffer.Sequence, session.Account);
+                        message = buffer.GetMessage();
+                        partialFragments.Remove(fragment.Header.Sequence);
                     }
                 }
+                else
+                {
+                    // No existing buffer, so add a new one for this fragment sequence.
+                    log.DebugFormat("[{0}] Creating new buffer {1} for this split fragment", session.Account, fragment.Header.Sequence);
+                    partialFragments.Add(fragment.Header.Sequence, new MessageBuffer(fragment.Header.Sequence, fragment.Header.Count));
+                }
             }
             else
             {
-                ClientMessage message = new ClientMessage(fragment.Data);
-                InboundMessageManager.HandleClientMessage(message, session);
+                // Packet is not split, proceed with handling it.
+                log.DebugFormat("[{0}] Fragment {1} is not split", session.Account, fragment.Header.Sequence);
+                message = new ClientMessage(fragment.Data);
+            }
+
+            // If message is not null, we have a complete message to handle
+            if (message != null)
+            {
+                // First check if this message is the next sequence, if it is not, add it to our outOfOrderFragments
+                if (fragment.Header.Sequence == lastReceivedFragmentSequence + 1)
+                {
+                    log.DebugFormat("[{0}] Handling fragment {1}", session.Account, fragment.Header.Sequence);
+                    HandleFragment(message);
+                }
+                else
+                {
+                    log.DebugFormat("[{0}] Fragment {1} is early, lastReceivedFragmentSequence = {2}", session.Account, fragment.Header.Sequence, lastReceivedFragmentSequence);
+                    // outOfOrderFragments is only ever touched in the main Update loop, so no need to lock.

```

Comment:
This is not true, need to lock this actually, as this can come in on ProcessPacket